File: backend/src/routes/reports.py
import csv
import io
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from fpdf import FPDF

from src.database import get_connection
from src.dependencies import get_authenticated_user

router = APIRouter(prefix="/api/reports", tags=["reports"])


def _get_client_and_entries(user_email: str, client_id: int):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id, name FROM clients WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            client = cur.fetchone()
            if not client:
                raise HTTPException(status_code=404, detail="Client not found")

            cur.execute(
                "SELECT id, hours, description, date, created_at, updated_at "
                "FROM work_entries WHERE client_id = %s AND user_email = %s "
                "ORDER BY date DESC",
                (client_id, user_email),
            )
            entries = cur.fetchall()
            return client, entries
    finally:
        conn.close()


@router.get("/client/{client_id}")
def get_client_report(client_id: int, user_email: str = Depends(get_authenticated_user)):
    client, entries = _get_client_and_entries(user_email, client_id)
    total_hours = sum(float(e["hours"]) for e in entries)
    return {
        "client": client,
        "workEntries": entries,
        "totalHours": total_hours,
        "entryCount": len(entries),
    }


@router.get("/export/csv/{client_id}")
def export_csv(client_id: int, user_email: str = Depends(get_authenticated_user)):
    client, entries = _get_client_and_entries(user_email, client_id)

    buf = io.StringIO()
    writer = csv.DictWriter(buf, fieldnames=["Date", "Hours", "Description", "Created At"])
    writer.writeheader()
    for e in entries:
        writer.writerow({
            "Date": str(e["date"]),
            "Hours": str(e["hours"]),
            "Description": e["description"] or "",
            "Created At": str(e["created_at"]),
        })

    safe_name = "".join(c if c.isalnum() else "_" for c in client["name"])
    timestamp = datetime.now(tz=timezone.utc).isoformat().replace(":", "-").replace(".", "-")
    filename = f"{safe_name}_report_{timestamp}.csv"

    buf.seek(0)
    return StreamingResponse(
        iter([buf.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get("/export/pdf/{client_id}")
def export_pdf(client_id: int, user_email: str = Depends(get_authenticated_user)):
    client, entries = _get_client_and_entries(user_email, client_id)
    total_hours = sum(float(e["hours"]) for e in entries)

    pdf = FPDF()
    pdf.add_page()

    pdf.set_font("Helvetica", "B", 20)
    pdf.cell(0, 10, f"Time Report for {client['name']}", align="C", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(5)

    pdf.set_font("Helvetica", "", 14)
    pdf.cell(0, 8, f"Total Hours: {total_hours:.2f}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 8, f"Total Entries: {len(entries)}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 8, f"Generated: {datetime.now(tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')}", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(5)

    pdf.set_font("Helvetica", "B", 12)
    pdf.cell(40, 8, "Date")
    pdf.cell(30, 8, "Hours")
    pdf.cell(0, 8, "Description", new_x="LMARGIN", new_y="NEXT")
    pdf.line(10, pdf.get_y(), 200, pdf.get_y())
    pdf.ln(2)

    pdf.set_font("Helvetica", "", 11)
    for i, entry in enumerate(entries):
        if pdf.get_y() > 270:
            pdf.add_page()
        y = pdf.get_y()
        pdf.cell(40, 7, str(entry["date"]))
        pdf.cell(30, 7, str(entry["hours"]))
        pdf.cell(0, 7, entry["description"] or "No description", new_x="LMARGIN", new_y="NEXT")
        if (i + 1) % 5 == 0:
            pdf.line(10, pdf.get_y(), 200, pdf.get_y())
            pdf.ln(2)

    safe_name = "".join(c if c.isalnum() else "_" for c in client["name"])
    timestamp = datetime.now(tz=timezone.utc).isoformat().replace(":", "-").replace(".", "-")
    filename = f"{safe_name}_report_{timestamp}.pdf"

    buf = io.BytesIO(pdf.output())
    buf.seek(0)
    return StreamingResponse(
        buf,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: backend/src/routes/reports.js
const express = require('express');
const { getDatabase } = require('../database/init');
const { authenticateUser } = require('../middleware/auth');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;
const PDFDocument = require('pdfkit');
const path = require('path');
const fs = require('fs');

const router = express.Router();

function entryAmount(entry, hourlyRate) {
  return entry.billable ? (Number.parseFloat(entry.hours) || 0) * hourlyRate : 0;
}

function roundMoney(amount) {
  return Math.round(amount * 100) / 100;
}

function fetchClientReportData(clientId, userEmail, callback) {
  const db = getDatabase();

  db.get(
    'SELECT id, name, hourly_rate AS hourlyRate FROM clients WHERE id = ? AND user_email = ?',
    [clientId, userEmail],
    (err, client) => {
      if (err || !client) {
        return callback(err, client);
      }

      const hourlyRate = Number.parseFloat(client.hourlyRate) || 0;

      db.all(
        `SELECT id, hours, description, date, billable, created_at, updated_at
         FROM work_entries 
         WHERE client_id = ? AND user_email = ? 
         ORDER BY date DESC`,
        [clientId, userEmail],
        (err, workEntries) => {
          if (err) {
            return callback(err);
          }

          callback(null, { client, hourlyRate, workEntries });
        }
      );
    }
  );
}

function handleClientReport(req, res, onData) {
  const clientId = Number.parseInt(req.params.clientId, 10);

  if (Number.isNaN(clientId)) {
    return res.status(400).json({ error: 'Invalid client ID' });
  }

  fetchClientReportData(clientId, req.userEmail, (err, data) => {
    if (err) {
      console.error('Database error:', err);
      return res.status(500).json({ error: 'Internal server error' });
    }

    if (!data) {
      return res.status(404).json({ error: 'Client not found' });
    }

    onData(data);
  });
}

// All routes require authentication
router.use(authenticateUser);

// Get hourly report for specific client
router.get('/client/:clientId', (req, res) => {
  handleClientReport(req, res, ({ client, hourlyRate, workEntries }) => {
    const workEntriesWithAmounts = workEntries.map(entry => ({
      ...entry,
      amount: roundMoney(entryAmount(entry, hourlyRate))
    }));

    const totalHours = workEntries.reduce((sum, entry) => sum + Number.parseFloat(entry.hours), 0);
    const totalAmount = roundMoney(
      workEntriesWithAmounts.reduce((sum, entry) => sum + entry.amount, 0)
    );

    res.json({
      client,
      workEntries: workEntriesWithAmounts,
      totalHours,
      totalAmount,
      entryCount: workEntries.length
    });
  });
});

// Export client report as CSV
router.get('/export/csv/:clientId', (req, res) => {
  handleClientReport(req, res, ({ client, hourlyRate, workEntries }) => {
    // Create temporary CSV file
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `${client.name.replace(/[^a-zA-Z0-9]/g, '_')}_report_${timestamp}.csv`;
    const tempPath = path.join(__dirname, '../../temp', filename);

    // Ensure temp directory exists
    const tempDir = path.dirname(tempPath);
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir, { recursive: true });
    }

    const csvWriter = createCsvWriter({
      path: tempPath,
      header: [
        { id: 'date', title: 'Date' },
        { id: 'hours', title: 'Hours' },
        { id: 'billable', title: 'Billable' },
        { id: 'amount', title: 'Amount' },
        { id: 'description', title: 'Description' },
        { id: 'created_at', title: 'Created At' }
      ]
    });

    const records = workEntries.map(entry => ({
      ...entry,
      billable: entry.billable ? 'Yes' : 'No',
      amount: roundMoney(entryAmount(entry, hourlyRate)).toFixed(2)
    }));

    csvWriter.writeRecords(records)
      .then(() => {
        // Send file and clean up
        res.download(tempPath, filename, (err) => {
          if (err) {
            console.error('Error sending file:', err);
          }
          // Clean up temp file
          fs.unlink(tempPath, (unlinkErr) => {
            if (unlinkErr) {
              console.error('Error deleting temp file:', unlinkErr);
            }
          });
        });
      })
      .catch((error) => {
        console.error('Error creating CSV:', error);
        res.status(500).json({ error: 'Failed to generate CSV report' });
      });
  });
});

// Export client report as PDF
router.get('/export/pdf/:clientId', (req, res) => {
  handleClientReport(req, res, ({ client, hourlyRate, workEntries }) => {
    // Create PDF
    const doc = new PDFDocument();
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `${client.name.replace(/[^a-zA-Z0-9]/g, '_')}_report_${timestamp}.pdf`;

    // Set response headers
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);

    // Pipe PDF to response
    doc.pipe(res);

    // Add content to PDF
    doc.fontSize(20).text(`Time Report for ${client.name}`, { align: 'center' });
    doc.moveDown();

    const totalHours = workEntries.reduce((sum, entry) => sum + Number.parseFloat(entry.hours), 0);
    const totalAmount = roundMoney(
      workEntries.reduce((sum, entry) => sum + entryAmount(entry, hourlyRate), 0)
    );
    doc.fontSize(14).text(`Total Hours: ${totalHours.toFixed(2)}`);
    doc.text(`Total Amount: ${totalAmount.toFixed(2)}`);
    doc.text(`Total Entries: ${workEntries.length}`);
    doc.text(`Generated: ${new Date().toLocaleString()}`);
    doc.moveDown();

    // Add table header
    doc.fontSize(12).text('Date', 50, doc.y, { width: 90 });
    doc.text('Hours', 140, doc.y - 15, { width: 60 });
    doc.text('Amount', 200, doc.y - 15, { width: 80 });
    doc.text('Description', 285, doc.y - 15, { width: 265 });
    doc.moveDown();

    // Add horizontal line
    doc.moveTo(50, doc.y).lineTo(550, doc.y).stroke();
    doc.moveDown(0.5);

    // Add work entries
    workEntries.forEach((entry, index) => {
      const y = doc.y;

      // Check if we need a new page
      if (y > 700) {
        doc.addPage();
      }

      doc.text(entry.date, 50, doc.y, { width: 90 });
      doc.text(entry.hours.toString(), 140, y, { width: 60 });
      doc.text(roundMoney(entryAmount(entry, hourlyRate)).toFixed(2), 200, y, { width: 80 });
      doc.text(entry.description || 'No description', 285, y, { width: 265 });
      doc.moveDown();

      // Add separator line every 5 entries
      if ((index + 1) % 5 === 0) {
        doc.moveTo(50, doc.y).lineTo(550, doc.y).stroke();
        doc.moveDown(0.5);
      }
    });

    // Finalize PDF
    doc.end();
  });
});

module.exports = router;
